README.md
Iquidus Explorer - 1.7.3.emd
================

An open source block explorer written in node.js.

Working with these EMD exchanges

*  [FreiExchange](https://freiexchange.com/market/EMD/BTC)
*  [Graviex Exchange](https://graviex.net/markets/emdbtc)
*  [Unnamed Exchange](https://www.unnamed.exchange/Exchange/Basic?market=EMD_BTC)

### See it in action

*  [List of live explorers running Iquidus](https://github.com/iquidus/explorer/wiki/Live-Explorers)


*Note: If you would like your instance mentioned here contact me*

### Requires

*  node.js >= 8.17.0 (12.14.0 is advised for updated dependencies)
*  mongodb 4.2.x
*  *coind

### Create database

Enter MongoDB cli:

    $ mongo

Create databse:

    > use explorerdb

Create user with read/write access:

    > db.createUser( { user: "iquidus", pwd: "3xp!0reR", roles: [ "readWrite" ] } )

*Note: If you're using mongo shell 4.2.x, use the following to create your user:

    > db.addUser( { user: "username", pwd: "password", roles: [ "readWrite"] })

### Get the source

    git clone https://github.com/Crypto-Currency/Emerald-Block-Explorer.git

### Install node modules

    cd Emerald-Block-Explorer && npm install --production

### Configure

    cp ./settings.json.template ./settings.json

*Make required changes in settings.json*

### Start Explorer

    npm start

*Note: mongod must be running to start the explorer*

As of version 1.4.0 the explorer defaults to cluster mode, forking an instance of its process to each cpu core. This results in increased performance and stability. Load balancing gets automatically taken care of and any instances that for some reason die, will be restarted automatically. For testing/development (or if you just wish to) a single instance can be launched with

    node --stack-size=10000 bin/instance

To stop the cluster you can use

    npm stop

### Syncing databases with the blockchain

sync.js (located in scripts/) is used for updating the local databases. This script must be called from the explorers root directory.

    Usage: node scripts/sync.js [database] [mode]

    database: (required)
    index [mode] Main index: coin info/stats, transactions & addresses
    market       Market data: summaries, orderbooks, trade history & chartdata

    mode: (required for index database only)
    update       Updates index from last sync to current block
    check        checks index for (and adds) any missing transactions/addresses
    reindex      Clears index then resyncs from genesis to current block

    notes:
    * 'current block' is the latest created block when script is executed.
    * The market database only supports (& defaults to) reindex mode.
    * If check mode finds missing data(ignoring new data since last sync),
      index_timeout in settings.json is set too low.


*It is recommended to have this script launched via a cronjob at 1+ min intervals.*

**crontab**

*Example crontab; update index every minute and market data every 2 minutes*

    */1 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js index update > /dev/null 2>&1
    */2 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js market > /dev/null 2>&1
    */5 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/peers.js > /dev/null 2>&1

### Wallet

Iquidus Explorer is intended to be generic, so it can be used with any wallet following the usual standards. The wallet must be running with atleast the following flags

    -daemon -txindex
    
### Security

Ensure mongodb is not exposed to the outside world via your mongo config or a firewall to prevent outside tampering of the indexed chain data. 

### Known Issues

**script is already running.**

If you receive this message when launching the sync script either a) a sync is currently in progress, or b) a previous sync was killed before it completed. If you are certian a sync is not in progress remove the index.pid and db_index.pid from the tmp folder in the explorer root directory.

    rm tmp/index.pid
    rm tmp/db_index.pid

**exceeding stack size**

    RangeError: Maximum call stack size exceeded

Nodes default stack size may be too small to index addresses with many tx's. If you experience the above error while running sync.js the stack size needs to be increased.

To determine the default setting run

    node --v8-options | grep -B0 -A1 stack_size

To run sync.js with a larger stack size launch with

    node --stack-size=[SIZE] scripts/sync.js index update

Where [SIZE] is an integer higher than the default.

*note: SIZE will depend on which blockchain you are using, you may need to play around a bit to find an optimal setting*

### License

Copyright (c) 2015, Iquidus Technology  
Copyright (c) 2015, Luke Williams  
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Iquidus Technology nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGELOG
1.7.3.emd
* adding Graviex Exchange market
* adding FreiExchange market
* adding Unnamed Exchange market

1.7.3
* Fixes for Historical Address Balance - seems to be fastest method that works along-side parallel indexing
* New settings to optionally show market cap in header: show_market_cap and show_market_cap_over_price

1.7.2
* Index blocks in parallel - uses settings for task counts.
* Indexing is much faster (again).
* New settings to adjust parallel task counts: block_parallel_tasks

1.7.1
* Added "claim address" system - allows users to label addresses by signing messages

1.7.0
* Various updated dependencies & node version
* Switched from Jade to Pug
* Updated required mongo version

1.6.2
* Indexing is now MUCH faster.
* Added AddressTX model/collection - REQUIRES REINDEX
* Added balance history to Address TX History
* Added support for AJAX loading of Address TX History, using txcount setting for loading
* Stopped maximum TX count trimming TX's in AddressTX collection
* Removed tx_array from Address collection -> AddressTX
* Fix mismatching balances/sent/received, negative balances etc
* Added new file lock during database indexing: tmp/db_index.pid
* New setting to lock during indexing: lock_during_index
* Add setting to call bitcoin-core directly or use RPC during indexing

1.6.1
* fixed last_txs setting
* added hashrate_units setting
* added cryptsy support (markets)
* removed market update lockfile

1.6.0
* Fixed negative/incorrect balance due to sending to yourself (Issues #6, #8)
* Pages found via search now display full url 
* Markets backend overhaul/rewrite
* Added Yobit support (markets)
* Added Empoex support (markets)
* Layout redesign
* Layout data (coin supply, connections etc) now fetched via ajax
* Dynamic table data on main view (ajax, updates every 30 seconds)

1.5.2
* fix poloniex href (xCore)
* added links (xCore)
* coinbase balance fix (xCore)
* getaddress balance fix
* added getbalance api call
* finalized getlasttxs api call
* added footer logo

1.5.1
* added labels setting (address labels)

1.5.0
* added theme setting to settings.json
* added bootswatch themes
* new extended api call (getlasttxs)
* Index page now shows last x transactions instead of only last block
* added last_txs settings to settings.json
* mobile fixes
* OP_RETURN/type=nulldata crash fix
* PoS fixes
* search tx confirmation fix
* search tx blockcount fix
* address page redesign

1.4.1

* Added user and password to database settings
* Mongo authentication 
* Added alert (index not up-to-date)
* New richlist configuration settings
* added 'npm stop' to kill cluster
* Added logo setting

1.4.0

* Removed mintpal support (markets)
* Added poloniex support (markets)
* Cluster support
* Jasmine test scripts
* getaddress api fix (address not found)
* Base 2 rounding fix (values stored in satoshis)
* DRK fixes
* New current supply setting: BALANCES
* Sync fixes/optimizations
* New setting: show_sent_received

1.3.4

* Added qrcode api
* Added qrcode to address page

1.3.3

* Wealth Distribution stats & charts (richlist)
* New locale settings (richlist)
* Extended api
* getmoneysupply (api)
* getaddress (api)
* getdistribution (api)
* Negative zero balance fix (address)

1.3.2

* HeavyCoin support
* New heavy setting (heavy)
* New rewards page (heavy)
* Current coin supply (coinbase total sent, getinfo.moneysupply, getsupply (heavy))
* New supply setting (supply)
* New txcount setting
* Address txs segfault fix
* Indexing optimizations
* DOS address page fix

1.3.1

* Rich Lists/Top 100
* PoS fixes
* ui fixes
* coinbase address segfault fix
* performance tweaks

1.3.0

* merged blocknotify.js and markets.js into sync.js
* added update and check timeout settings
* local tx index
* input addresses/amounts
* unixtime format (now human friendly date/time)
* fixed genesis crash
* added genesis settings
* local address index
* new address view
* search by address

1.2.1

* updated locale (markets)
* updated index view
* updated tx view
* updated block view
* added optional seach to navbar
* added confirmations setting
* added screenshots

1.2.0

* added jqplot and datatables
* new markets page 
* added mintpal support (markets)
* added bittrex support (markets)
* new display setting (markets, twitter) 
* new locale entries (markets) 
* new settings (markets, api)
* added icons to nav (explorer, markets, api, loading, twitter)
* new twitter setting

1.1.0

* fixed navbar title (was not reading from settings)
* added locale support (and updated views)
* new display setting to toggle menu items (api)

1.0.0

* initial release


lib/locale.js
/**
* The Locale Module reads the locale settings and provides
* this information to the other modules
*/

var fs = require("fs");
var jsonminify = require("jsonminify");
var settings = require("./settings");

exports.menu_explorer = "Explorer",
exports.menu_api = "API",
exports.menu_markets = "Markets",
exports.menu_richlist = "Rich List",
exports.menu_reward = "Reward",
exports.menu_movement = "Movement",
exports.menu_node = "Nodes",
exports.menu_network = "Network"

exports.ex_title = "Block Explorer",
exports.ex_search_title = "Search",
exports.ex_search_button = "Search",
exports.ex_search_message = "You may enter a block height, block hash ,tx hash or address.",
exports.ex_error = "Error!",
exports.ex_warning = "Warning:",
exports.ex_search_error = "Search found no results.",
exports.ex_latest_transactions = "Latest Transactions",
exports.ex_summary = "Block Summary",
exports.ex_supply = "Coin Supply",
exports.ex_block = "Block",
exports.tx_title = "Transaction Details",
exports.tx_block_hash = "Block Hash",
exports.tx_recipients = "Recipients",
exports.tx_contributors = "Contributor(s)",
exports.tx_hash = "Hash",
exports.tx_address = "Address",
exports.tx_nonstandard = "NONSTANDARD TX",

exports.block_title = "Block Details",
exports.block_previous = "Previous",
exports.block_next = "Next",
exports.block_genesis = "GENESIS",

exports.difficulty = "Difficulty",
exports.network = "Network",
exports.height = "Height",
exports.timestamp = "Timestamp",
exports.size = "Size",
exports.transactions = "Transactions",
exports.total_sent = "Total Sent",
exports.total_received = "Total Received",
exports.confirmations = "Confirmations",
exports.total = "Total",
exports.bits = "Bits",
exports.nonce = "Nonce",
exports.new_coins = "New Coins",
exports.proof_of_stake = "PoS",
exports.initial_index_alert = "Indexing is currently incomplete, functionality is limited until index is up-to-date.",

exports.a_menu_showing = "Showing",
exports.a_menu_txs = "transactions",
exports.a_menu_all = "All",
exports.a_qr = "QR Code",

exports.rl_received_coins = "Top 100 - Received Coins",
exports.rl_current_balance = "Top 100 - Current Balance",
exports.rl_received = "Received",
exports.rl_balance = "Balance",
exports.rl_wealth = "Wealth Distribution",
exports.rl_top25 = "Top 1-25",
exports.rl_top50 = "Top 26-50",
exports.rl_top75 = "Top 51-75",
exports.rl_top100 = "Top 76-100",
exports.rl_hundredplus = "101+",

exports.net_addnodes = "Add Nodes",
exports.net_connections = "Connections",
exports.net_address = "Address",
exports.net_protocol = "Protocol",
exports.net_subversion = "Sub-version",
exports.net_country = "Country",
exports.net_warning = "This is simply a sub sample of the network based on wallets connected to this node.",

exports.api_title = "API Documentation",
exports.api_message = "The block explorer provides an API allowing users and/or applications to retrieve information from the network without the need for a local wallet.",
exports.api_calls = "API Calls",
exports.api_getnetworkhashps = "Returns the current network hashrate. (hash/s)",
exports.api_getdifficulty = "Returns the current difficulty.",
exports.api_getconnectioncount = "Returns the number of connections the block explorer has to other nodes.",
exports.api_getblockcount = "Returns the number of blocks currently in the block chain.",
exports.api_getblockhash = "Returns the hash of the block at ; index 0 is the genesis block.",
exports.api_getblock = "Returns information about the block with the given hash.",
exports.api_getrawtransaction = "Returns raw transaction representation for given transaction id. decrypt can be set to 0(false) or 1(true).",
exports.api_getmaxmoney = 'Returns the maximum possible money supply.',
exports.api_getmaxvote = 'Returns the maximum allowed vote for the current phase of voting.',
exports.api_getvote = 'Returns the current block reward vote setting.',
exports.api_getphase = 'Returns the current voting phase (\'Mint\', \'Limit\' or \'Sustain\').',
exports.api_getreward = 'Returns the current block reward, which has been decided democratically in the previous round of block reward voting.',
exports.api_getsupply = 'Returns the current money supply.',
exports.api_getnextrewardestimate = 'Returns an estimate for the next block reward based on the current state of decentralized voting.',
exports.api_getnextrewardwhenstr =  'Returns string describing how long until the votes are tallied and the next block reward is computed.',

// Markets view
exports.mkt_hours = "24 hours",
exports.mkt_view_chart = "View 24 hour summary",
exports.mkt_view_summary = "View 24 hour chart",
exports.mkt_no_chart = "Chart data is not available via markets API.",
exports.mkt_high = "High",
exports.mkt_low = "Low",
exports.mkt_volume = "Volume",
exports.mkt_top_bid = "Top Bid",
exports.mkt_top_ask = "Top Ask",
exports.mkt_last = "Last Price",
exports.mkt_yesterday = "Yesterday",
exports.mkt_change = "Change",
exports.mkt_sell_orders = "Sell Orders",
exports.mkt_buy_orders = "Buy Orders",
exports.mkt_price = "Price",
exports.mkt_amount = "Amount",
exports.mkt_total = "Total",
exports.mkt_trade_history = "Trade History",
exports.mkt_type = "Type",
exports.mkt_time_stamp = "Time Stamp",
// Heavy

exports.heavy_vote = "Vote",
    // Heavy rewards view
exports.heavy_title = "Reward/voting information",

exports.heavy_cap = "Coin Cap",
exports.heavy_phase = "Phase",
exports.heavy_maxvote = "Max Vote",
exports.heavy_reward = "Reward",
exports.heavy_current = "Current Reward",
exports.heavy_estnext = "Est. Next",
exports.heavy_changein = "Reward change in approximately",
exports.heavy_key = "Key",
exports.heavy_lastxvotes = "Last 20 votes",

exports.fides = "Fides-ex",
exports.poloniex = "Poloniex",
exports.bittrex = "Bittrex",
exports.altmarkets = "AltMarkets",
exports.bleutrade = "Bleutrade",
exports.yobit = "Yobit",
exports.cryptsy = "Cryptsy",
exports.empoex = "Empoex",
exports.ccex = "C-Cex",
exports.crex = "Crex24",
exports.tradesatoshi = "TradeSatoshi",
exports.freiexchange = "FreiExchange",
exports.graviex = "Graviex",
exports.unnamed = "Unnamed",

exports.reloadLocale = function reloadLocale(locale) {
  // Discover where the locale file lives
  var localeFilename = locale;
  //console.log(localeFilename);
  localeFilename = "./" + localeFilename;
  //console.log('Loading locale: ' + localeFilename);
  var localeStr;
  try{
    //read the settings sync
    localeStr = fs.readFileSync(localeFilename).toString();
  } catch(e){
    console.warn('Locale file not found. Continuing using defaults!');
  }

  // try to parse the settings
  var lsettings;
  try {
    if(localeStr) {
      localeStr = jsonminify(localeStr).replace(",]","]").replace(",}","}");
      lsettings = JSON.parse(localeStr);
    }
  }catch(e){
    console.error('There was an error processing your locale file: '+e.message);
    process.exit(1);
  }

  //loop trough the settings
  for(var i in lsettings)
  {
    //test if the setting start with a low character
    if(i.charAt(0).search("[a-z]") !== 0)
    {
      console.warn("Settings should start with a low character: '" + i + "'");
    }

    //we know this setting, so we overwrite it
    if(exports[i] !== undefined)
    {
      exports[i] = lsettings[i];
    }
    //this setting is unkown, output a warning and throw it away
    else
    {
      console.warn("Unknown Setting: '" + i + "'. This setting doesn't exist or it was removed");
    }
  }

};

// initially load settings
exports.reloadLocale(settings.locale);


lib/markets/altilly.js
var request = require('request');

var base_url = 'https://api.altilly.com/api/public';

function get_summary(coin, exchange, buys, sells, cb) {
  var summary = {};
  var topbid = 0;
  var topask = 0;
  //var req_url = base_url + '/ticker/' + coin + exchange;
  var req_url = base_url + '/ticker/EMDETH';
  request({uri: req_url, json: true}, function (error, response, body) {
        if (error) {
            return cb(error, null);
        } else if (response.statusCode === 200) {
            summary['market'] = "Altilly";
            summary['bid'] = parseFloat(body['bid']).toFixed(8);
            summary['ask'] = parseFloat(body['ask']).toFixed(8);
            summary['volume'] = parseFloat(body['volumeQuote']).toFixed(8);
            summary['high'] = parseFloat(body['high']).toFixed(8);
            summary['low'] = parseFloat(body['low']).toFixed(8);
            summary['last'] = parseFloat(body['last']).toFixed(8);
            summary['change'] = "0";

            return cb(null, summary);
        } else {
            return cb(error, null);
            return cb(error, "nully");
        }
  });
}

function get_trades(coin, exchange, cb) {
  var req_url = base_url + '/trades/' + coin + exchange + '?sort=DESC&by=timestamp&limit=100';
  request({uri: req_url, json: true}, function (error, response, body) {
    if (response.statusCode == 200) {
      var trades = [];
      if (body.length > 0) {
          for (var i = 0; i < body.length; i++) {
            var time = body[i]['unixtimestamp'];
            var trade = {
              type: body[i]['side'],
              price: body[i]['price'],
              amount: body[i]['quantity'],
              total: parseFloat((body[i]['quantity']) * (body[i]['price'])).toFixed(8),
              timestamp: time,
              id: body[i]['id']
            }
            trades.push(trade);
          }
      }
      trades = trades.sort(function(a,b){
        var datea = a.timestamp, dateb = b.timestamp;
        return dateb - datea;
      });
      return cb (null, trades);
    } else {
      return cb(body.message, null);
    }
  });
}

function get_orders(coin, exchange, cb) {
  var req_url = base_url + '/orderbook/' + coin + exchange + '?limit=100';
  request({uri: req_url, json: true}, function (error, response, body) {
    if (response.statusCode == 200) {
      var orders = body;
      var buys = [];
      var sells = [];
      if (orders['bid'].length > 0){
          for (var i = 0; i < orders['bid'].length; i++) {
            var order = {
              amount: parseFloat(orders.bid[i].size).toFixed(10),
              price: parseFloat(orders.bid[i].price).toFixed(10),
              //  total: parseFloat(orders.buy[i].Total).toFixed(8)
              // Necessary because API will return 0.00 for small volume transactions
              total: parseFloat(orders.bid[i].totsize).toFixed(10)
            }
            buys.push(order);
          }
      }
      if (orders['ask'].length > 0) {
        for (var x = 0; x < orders['ask'].length; x++) {
            var order = {
                amount: parseFloat(orders.ask[x].size).toFixed(10),
                price: parseFloat(orders.ask[x].price).toFixed(10),
                //    total: parseFloat(orders.sell[x].Total).toFixed(8)
                // Necessary because API will return 0.00 for small volume transactions
                total: parseFloat(orders.ask[x].totsize).toFixed(10)
            }
            sells.push(order);
        }
      }
      return cb(null, buys, sells);
    } else {
      return cb(body.message, [], []);
    }
  });
}

module.exports = {
  get_data: function(coin, exchange, cb) {
    var error = null;
    get_orders(coin, exchange, function(err, buys, sells) {
      if (err) { error = err; }
      get_trades(coin, exchange, function(err, trades) {
        if (err) { error = err; }
        get_summary(coin, exchange, buys, sells, function(err, stats) {
          if (err) { error = err; }
          return cb(error, {buys: buys, sells: sells, chartdata: [], trades: trades, stats: stats});
        });
      });
    });
  }
};


lib/markets/graviex.js
var request = require('request');

var base_url = 'https://graviex.net/webapi/v3';

function get_summary(coin, exchange, cb) {
  var summary = {};
  var url=(base_url + '/tickers/' + coin.concat(exchange)).toLowerCase();
  request({"rejectUnauthorized": false, uri: url, json: true}, function (error, response, body) {
    if (error) {
      return cb(error, null);
    } else if (body.error !== true) {

      summary['bid'] = parseFloat(body['buy']).toFixed(9);
      summary['ask'] = parseFloat(body['sell']).toFixed(9);
      request({"rejectUnauthorized": false, uri: url , json: true}, function (error, response, body) {
        if (error) {
          return cb(error, null);
        } else if (body.error !== true) {
          summary['volume'] = parseFloat(body['volume']);
          summary['volume_btc'] = parseFloat(body['volume2']);
          summary['high'] = parseFloat(body['high']).toFixed(9);
          summary['low'] = parseFloat(body['low']).toFixed(9);
          summary['last'] = parseFloat(body['last']).toFixed(9);
          return cb(null, summary);
        } else {
          return cb(error, null);      
        }
      });
    } else {
      return cb(error, null);
    }
  });   
}
function get_trades(coin, exchange, cb) {
	  var req_url = (base_url + '/trades_simple.json?market=' + coin.concat(exchange)).toLowerCase();
  request({ uri: req_url, json: true }, function (error, response, body) {
        if(error)
          return cb(error, null);
        else if (body.error !== true) {
          var tTrades = body;
          var trades = [];
          for (var i = 0; i < tTrades.length; i++) {
              var Trade = {
                  ordertype: tTrades[i].type.charAt(0).toUpperCase() + tTrades[i].type.substring(1).toLowerCase(),
                  amount: parseFloat(tTrades[i].amount).toFixed(4),
                  price: parseFloat(tTrades[i].price).toFixed(9),
                  total: (parseFloat(tTrades[i].amount).toFixed(4) * parseFloat(tTrades[i].price)).toFixed(9),
                  timestamp: formatTime(tTrades[i].date)
              }
              trades.push(Trade);
          }
	          return cb(null, trades);
      } else {
          return cb(body.Message, null);
      }
  });
}

function formatTime(timestamp){
    var raw = new Date(timestamp*1000);
    var months = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
    var year = raw.getFullYear();
    var month = months[raw.getMonth()];
    var date = ("0" +  raw.getDate()).substr(-2);	
    var hour = ("0" + raw.getHours()).substr(-2);
    var min = ("0" + raw.getMinutes()).substr(-2);
    var sec = ("0" + raw.getSeconds()).substr(-2);
    var time = date + ' ' + month + ' ' + year + ' ' + hour + ':' + min + ':' + sec ;
    return time;
}

function get_orders(coin, exchange, cb) {
        var req_url = (base_url + '/depth.json?market=' + coin.concat(exchange)).toLowerCase();
	request({ uri: req_url, json: true }, function (error, response, body) {
        if(error)
            return cb(error, null);
        else if (body.error !== true) {
            var buyorders = body['bids'];
            var sellorders = body['asks'];

            var buys = [];
            var sells = [];
            if (buyorders.length > 0){
                for (var i = 0; i < buyorders.length; i++) {
                    var order = {
                        amount: parseFloat(buyorders[i]['1']).toFixed(4),
                        price: parseFloat(buyorders[i]['0']).toFixed(9),
                        total: parseFloat(buyorders[i]['1']).toFixed(4) * parseFloat(buyorders[i]['0']).toFixed(9)
                    }
                   buys.push(order);
                  }
                  } else {}
                if (sellorders.length > 0) {
                for (var x = 0; x < sellorders.length; x++) {
                    var order = {
                        amount: parseFloat(sellorders[x]['1']).toFixed(9),
                        price: parseFloat(sellorders[x]['0']).toFixed(9),
                        total: parseFloat(sellorders[x]['1']).toFixed(9) * parseFloat(sellorders[x]['0']).toFixed(9)
                    }
                    sells.push(order);
                }
            } else {
            }
            return cb(null, buys, sells);
            } else {
            return cb(body.Message, [], [])
        }
     });
}

module.exports = {
  get_data: function(coin, exchange, cb) {
    var error = null;
     get_orders(coin, exchange, function(err, buys, sells) {
     if (err) { error = err; }
      get_trades(coin, exchange, function(err, trades) {
        if (err) { error = err; }
        get_summary(coin, exchange,  function(err, stats) {
          if (err) { error = err; }
          return cb(error, {buys: buys, sells: sells, chartdata: [], trades: trades, stats: stats});
        });
      });
    });
  }
};


lib/settings.js
/**
* The Settings Module reads the settings out of settings.json and provides
* this information to the other modules
*/

var fs = require("fs");
var jsonminify = require("jsonminify");


//The app title, visible e.g. in the browser window
exports.title = "blockchain";

//The url it will be accessed from
exports.address = "explorer.example.com";

// logo
exports.logo = "/images/logo.png";


//The app favicon fully specified url, visible e.g. in the browser window
exports.favicon = "favicon.ico";

//Theme
exports.theme = "Cyborg";

//The Port ep-lite should listen to
exports.port = process.env.PORT || 3001;


//coin symbol, visible e.g. MAX, LTC, HVC
exports.symbol = "BTC";


//coin name, visible e.g. in the browser window
exports.coin = "Bitcoin";


//This setting is passed to MongoDB to set up the database
exports.dbsettings = {
  "user": "iquidus",
  "password": "3xp!0reR",
  "database": "blockchaindb",
  "address" : "localhost",
  "port" : 27017
};


//This setting is passed to the wallet
exports.wallet = { "host" : "127.0.0.1",
  "port" : 8669,
  "user" : "bitcoinrpc",
  "pass" : "password"
};


//Locale file
exports.locale = "locale/en.json",


//Menu items to display
exports.display = {
  "api": true,
  "market": true,
  "twitter": true,
  "facebook": false,
  "googleplus": false,
  "youtube": false,
  "search": true,
  "richlist": true,
  "movement": true,
  "network": true
};


//API view
exports.api = {
  "blockindex": 1337,
  "blockhash": "00000000002db22bd47bd7440fcad99b4af5f3261b7e6bd23b7be911e98724f7",
  "txhash": "c251b0f894193dd55664037cbf4a11fcd018ae3796697b79f5097570d7de95ae",
  "address": "RBiXWscC63Jdn1GfDtRj8hgv4Q6Zppvpwb",
};

// markets
exports.markets = {
  "coin": "EMD",
  "exchange": "BTC",
  "enabled": ['freiexchange'],
  "default": "freiexchange"
};

// richlist/top100 settings
exports.richlist = {
  "distribution": true,
  "received": true,
  "balance": true
};

exports.movement = {
  "min_amount": 100,
  "low_flag": 1000,
  "high_flag": 10000
},

//index
exports.index = {
  "show_hashrate": false,
  "difficulty": "POW",
  "last_txs": 100
};

// twitter
exports.twitter = "iquidus";
exports.facebook = "yourfacebookpage";
exports.googleplus = "yourgooglepluspage";
exports.youtube = "youryoutubechannel";

exports.confirmations = 6;

//timeouts
exports.update_timeout = 125;
exports.check_timeout = 250;


//genesis
exports.genesis_tx = "65f705d2f385dc85763a317b3ec000063003d6b039546af5d8195a5ec27ae410";
exports.genesis_block = "b2926a56ca64e0cd2430347e383f63ad7092f406088b9b86d6d68c2a34baef51";

exports.heavy = false;
exports.txcount = 100;
exports.show_sent_received = true;
exports.supply = "COINBASE";
exports.nethash = "getnetworkhashps";
exports.nethash_units = "G";

exports.labels = {};

exports.reloadSettings = function reloadSettings() {
  // Discover where the settings file lives
  var settingsFilename = "settings.json";
  settingsFilename = "./" + settingsFilename;

  var settingsStr;
  try{
    //read the settings sync
    settingsStr = fs.readFileSync(settingsFilename).toString();
  } catch(e){
    console.warn('No settings file found. Continuing using defaults!');
  }

  // try to parse the settings
  var settings;
  try {
    if(settingsStr) {
      settingsStr = jsonminify(settingsStr).replace(",]","]").replace(",}","}");
      settings = JSON.parse(settingsStr);
    }
  }catch(e){
    console.error('There was an error processing your settings.json file: '+e.message);
    process.exit(1);
  }

  //loop trough the settings
  for(var i in settings)
  {
    //test if the setting start with a low character
    if(i.charAt(0).search("[a-z]") !== 0)
    {
      console.warn("Settings should start with a low character: '" + i + "'");
    }

    //we know this setting, so we overwrite it
    if(exports[i] !== undefined)
    {
      exports[i] = settings[i];
    }
    //this setting is unkown, output a warning and throw it away
    else
    {
      console.warn("Unknown Setting: '" + i + "'. This setting doesn't exist or it was removed");
    }
  }

};

// initially load settings
exports.reloadSettings();
